### DISCUSSION TITLE
How to get pending/queued jobs?

### DISCUSSION THREAD

**Discussion by koskit** [quartznet/quartznet#2071] [Q&A] [answered]
Hello!

I have an implementation which consists of large jobs (think in 50-60 minutes each job). Basically I do some data manipulation on databases based on when I have an available backup from my sql server (the job reads the directory and if it has found new backup, it deletes some stuff, shrinks it etc.)

I have this schedule to run each hour.

Now, if a job has started, I don't want another one to start (since it uses large resources on the machine, RAM/CPU etc).

I have set the following:
`q.UseDefaultThreadPool(tp =>
                {
                    tp.MaxConcurrency = 1;
                });`
when registering the scheduler on microsoft's dependency injection, so I'm fine with blocking concurrent executions.

The thing is, if the execution takes more time (up to 2-3 hours), then more jobs will be fired and I will have somewhat of a "queue". 

I tested that (with a temporary Thread.Sleep()) and I saw that indeed, after the first one finished, the second starts even if it's delayed by 2-3 minutes (got fired but because of MaxConcurrency = 1, started later).

My question is, is there a way to get such jobs? That their schedule is past and have fired, but not started yet?

Edit: By "get", I mean something like Scheduler.GetPendingJobs(), in order to have a list on what is delayed because of the current job.

**Comment by lahma** [comment]
I'll keep posting this FAQ entry I guess https://www.quartz-scheduler.net/documentation/faq.html#how-do-i-keep-a-job-from-firing-concurrently

**Reply by koskit** [reply]
Please take a look at the question at the bottom of my answer my friend, I did read the wiki and managed to set the concurrency to what I wanted, my question isn't that

**Comment by koskit** [comment, marked as answer]
For anyone else searching for the same thing (getting the jobs that have pending execution but are blocked because of 
`[DisallowConcurrentExecution]`), I managed to get them with reflection on the job store with the following code (it only works for RAMJobStore, but could be altered to work for other types of job store I suppose.

    private static bool TryGetBlockedJobsFromScheduler(IScheduler scheduler, out List<JobKey> blockedJobs)
    {
        blockedJobs = new();

        BindingFlags bindingFlags = BindingFlags.Instance | BindingFlags.NonPublic;

        if (scheduler is not StdScheduler std)
            return false;

        QuartzScheduler? sched = typeof(StdScheduler).GetField("sched", bindingFlags)
                                                    ?.GetValue(std) as QuartzScheduler;

        if (sched == null)
            return false;

        QuartzSchedulerResources? resources = typeof(QuartzScheduler).GetField("resources", bindingFlags)
                                                                    ?.GetValue(sched) as QuartzSchedulerResources;

        if (resources == null)
            return false;

        if (resources.JobStore is RAMJobStore ramJobStore)
        {
            HashSet<JobKey>? blockedJobsInternal =
                typeof(RAMJobStore).GetField("blockedJobs", bindingFlags)
                                  ?.GetValue(ramJobStore) as HashSet<JobKey>;

            if (blockedJobsInternal != null)
            {
                blockedJobs = blockedJobsInternal.ToList();
                return true;
            }
        }

        return false;
    }
